lookup join first draft (#123719)
elastic/elasticsearch: docs/reference/query-languages/esql/esql-lookup-join.md

---
navigation_title: "Correlate data with LOOKUP JOIN"
mapped_pages:
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-enrich-data.html
---

# LOOKUP JOIN [esql-lookup-join]

> The {{esql}} [`LOOKUP join`](/reference/query-languages/esql/esql-commands.md#esql-lookup-join) processing command combines, at query-time, data from one or more source indexes with field-value combinations found in an input table. Teams often have data scattered across multiple indices – like logs, IPs, user IDs, hosts, employees etc. Without a direct way to enrich or correlate each event with reference data, root-cause analysis, security checks, and operational insights become time-consuming.

Reviewer: nit: I don't understand what "field-value combinations" exactly means.

For example, you can use `LOOKUP JOIN` to:

* Pull in environment or ownership details for each host to correlate your metrics data.
* Quickly see if any source IPs match known malicious addresses.
* Tag logs with the owning team or escalation info for faster triage and incident response.

[`LOOKUP join`](/reference/query-languages/esql/esql-commands.md#esql-lookup-join) is similar to [`ENRICH`](/reference/query-languages/esql/esql-commands.md#esql-enrich) in the fact that they both help you join data together. You should use `LOOKUP JOIN` when:

* Enrichment data changes frequently
* You want to avoid index time processing
* Working with regular indices
* Need to preserve distinct matches

> * Need to match on any field in a lookup index

Reviewer: The security considerations apply likewise here, if we want to add them (we should somewhere, right?)

## How the `LOOKUP JOIN` command works [esql-how-lookup-join-works]

The `LOOKUP JOIN` command adds new columns to a table, with data from {{es}} indices.

:::{image} ../../../images/esql-lookup-join.png
:alt: esql lookup join
:::

$$$esql-source-index$$$

Source index

> :   An index which stores enrich data that the `LOOKUP` command can add to input tables. You can create and manage these indices just like a regular {{es}} index. You can use multiple source indices in an enrich policy. You also can use the same source index in multiple enrich policies.

Reviewer: ```suggestion
:   An index which stores data that the `LOOKUP` command can add to input tables. You can create and manage these indices just like a regular {{es}} index. You also can use the same source index in multiple lookup joins.
```
Currently, only 1 lookup index can be used per `LOOKUP JOIN` command. Additionally, it cannot be a data stream, alias, or even index pattern. Date math is not allowed.

We plan to lift some of these limitations later (to be determined before GA, [see issue](https://github.com/elastic/elasticsearch/issues/120655)).

## Example

In the case where there are multiple matches on the index `LOOKUP JOIN` the output rows is the combination of each match from the left with each match on the right. 

> Image you have the two tables:

Reviewer: ```suggestion
Imagine you have the two tables:
```

**Left**

|Key|Entry|
| --- | --- |
|1|A|
|2|B|

**Right**

|Key|Language|TLD|
|---|---|---|
|1|English|CA|
|1|English|null|
|1|null|UK|
|1|English|US|

> |2|German|AT,DE|

Reviewer: bikeshed: this is some diiiirty data (that I created to be dirty, intentionally). Maybe there are more intuitive examples.

|2|German|CH|
|2|German|null|

Running the following query would provide the results shown below.

> ```esql

Reviewer: nit: referencing csv-tests would prevent this from going stale, but this is probably something for a follow-up in the future.

FROM Left
| WHERE Language IS NOT NULL // works and filter TLD UK
| LOOKUP JOIN Right ON Key
```

|Key|Entry|Language|TLD|
|---|---|---|---|
|1|A|English|CA|
|1|A|English|null|
|1|A|null|UK|
|1|A|English|US|
|2|A|German|AT,DE|
|2|B|German|CH|
|2|B|German|null|

::::{tip}

> `LOOKUP JOIN` does not guarantee the output to be in any particular order. If a certain order is required, users should use a [`SORT`](/reference/query-languages/esql/esql-commands.md#esql-sort) somewhere after the `LOOKUP JOIN`.

Reviewer: ++ this is important. We might provide _some_ guarantees later. There's an issue where the [sort order matters](https://github.com/elastic/elasticsearch/issues/121884#issuecomment-2643056468) (guarantees to be determined pre-GA).

::::

## Prerequisites [esql-enrich-prereqs]

To use `LOOKUP JOIN`, you must have:

* Data types of join key and join field in the lookup index need to generally be the same - up to widening of data types, where e.g. `short,byte` are considered equal to `integer`. Also, text fields can be used on the left hand side if and only if there is an exact subfield whose name is suffixed with `.keyword`.

## Limitations

The following are the current limitations with `LOOKUP JOIN`

> * `LOOKUP JOIN` will be sucessfull if both left and right type of the join are both `KEYWORD` types or if the left type is of `TEXT` and the right type is `KEYWORD`.

Reviewer: This is only true if the left hand side `TEXT` field has an exact `.keyword` subfield.

nit: the notions of "left" and "right" are not really defined for LOOKUP JOIN, I believe.

Reviewer: On second read, this looks maybe a bit misplaced? This is mentioned in more detail in the prerequisites above.

* Indices in [lookup](/reference/elasticsearch/index-settings/index-modules.md#index-mode-setting) mode are always single-sharded.
* Cross cluster search is unsupported. Both source and lookup indicies must be local.

> * `LOOKUP JOIN` can only use a single match field, and can only use a single index. Wildcards, aliases, and datastreams are not supported.

Reviewer: + datemath, also not supported

> * The name of the match field in `LOOKUP JOIN lu_idx ON match_field` must match an existing field in the query. This may require renames or evals to achieve it.

Reviewer: nit/bikeshed:
```suggestion
* The name of the match field in `LOOKUP JOIN lu_idx ON match_field` must match an existing field in the query. This may require renames or evals to achieve.
```

> * The query will circuit break if you fetch too much data in a single page. A large heap is needed to manage results of multiple megabytes.

Reviewer: I think it's better to drop the "single page" and say something along the lines of
```suggestion
* The query will circuit break if many documents from the lookup index have the same key. A large heap is needed to manage results of multiple megabytes per key.
```
@nik9000 , could you please keep me honest here?

>   * This limit is per page of data which is about about 10,000 rows.

Reviewer: I don't know if pages of data are a well-defined concept (at least user-facing). I also don't know if we can say this is about 10k rows. (@nik9000 may have more precise ideas on this.)

Author: It would be good to know the exacts here, I took a stab based on what I read but unsure what the limit is. Also will they get a specific messagE?

Reviewer: I'll check in with Nik to get a better, but still precise wording here.

Reviewer: Also, I don't think they will get a specific error message, probably just a generic circuit breaker exception.

Reviewer: Suggestion:
```
The query will circuit break if there are too many matching documents in the lookup index, or if the documents are too large. More precisely, `LOOKUP JOIN` works in batches of, normally, about 10,000 rows; a large amount of heap space is needed if the matching documents from the lookup index for a batch are multiple megabytes or larger. This is roughly the same as for `ENRICH`.
```
@nik9000 , could you please keep me honest here?

## Example

In the case where there are multiple matches on the index `LOOKUP JOIN` the output rows is the combination of each match from the left with each match on the right. 

Imagine you have the two tables:

**Left**

|Key|Entry|
| --- | --- |
|1|A|
|2|B|

**Right**

|Key|Language|TLD|
|---|---|---|
|1|English|CA|
|1|English|null|
|1|null|UK|
|1|English|US|
|2|German|AT,DE|
|2|German|CH|
|2|German|null|

Running the following query would provide the results shown below.

```esql
FROM employees
| WHERE Language IS NOT NULL // works and filter TLD UK
| LOOKUP JOIN Right ON Key
```

|Key|Entry|Language|TLD|
|---|---|---|---|
|1|A|English|CA|
|1|A|English|null|
|1|A|null|UK|
|1|A|English|US|
|2|A|German|AT,DE|
|2|B|German|CH|
|2|B|German|null|

::::{important}
`LOOKUP JOIN` does not guarantee the output to be in any particular order. If a certain order is required, users should use a [`SORT`](/reference/query-languages/esql/esql-commands.md#esql-sort) somewhere after the `LOOKUP JOIN`.

::::

> ## Prerequisites [esql-enrich-prereqs]

Reviewer 2: ```suggestion
## Prerequisites [esql-lookup-join-prereqs]
```

To use `LOOKUP JOIN`, you must have:

* Data types of join key and join field in the lookup index need to generally be the same. As long as the field data type is structurally capable of handling the data that you are passing through. For example `short,byte` is considered equal to `integer`. Also, text fields can be used on the left hand side if and only if there is an exact subfield whose name is suffixed with `.keyword`.

## Limitations

The following are the current limitations with `LOOKUP JOIN`

> * `LOOKUP JOIN` will be sucessfull if both left and right type of the join are both `KEYWORD` types or if the left type is of `TEXT` and the right type is `KEYWORD`.

Reviewer: This is incorrect, I think. It really is the main index that can have `keyword` fields; the same is not true for the lookup index.

> * `LOOKUP JOIN` will be sucessfull if both left and right type of the join are both `KEYWORD` types or if the left type is of `TEXT` and the right type is `KEYWORD`.

Reviewer 2: we definitely say subfield when talking about text and keyword, maybe "exact" is superfluous

---
navigation_title: "LOOKUP JOIN"
mapped_pages:
  - https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-enrich-data.html
---

# LOOKUP JOIN [esql-lookup-join]

The {{esql}} [`LOOKUP join`](/reference/query-languages/esql/esql-commands.md#esql-lookup-join) processing command combines, at query-time, data from one or more source indexes with field-value combinations found in an input table.

For example, you can use `LOOKUP JOIN` to:

> * Pull in environment or ownership details for each host to enrich your metrics data

Reviewer 2: 🚗 "augment" maybe?

* Quickly see if any source IPs match known malicious addresses.
* Tag logs with the owning team or escalation info for faster triage and incident response.

### How the `LOOKUP JOIN` command works [esql-how-lookup-join-works]

The `LOOKUP JOIN` command adds new columns to a table, with data from {{es}} indices. It requires a few special components:

:::{image} ../../../images/esql-lookup-join.png
:alt: esql lookup join
:::

$$$esql-source-index$$$

Source index
:   An index which stores enrich data that the `LOOKUP` command can add to input tables. You can create and manage these indices just like a regular {{es}} index. You can use multiple source indices in an enrich policy. You also can use the same source index in multiple enrich policies.

### Prerequisites [esql-enrich-prereqs]

To use `LOOKUP JOIN`, you must have:

> * Data types of join key and join field in the lookup index need to generally be the same - up to widening of data types, where e.g. `short,byte` are considered equal to `integer`. Also, text fields can be used on the left hand side if and only if there is an exact subfield whose name is suffixed with `.keyword`.

Author: I have added more verbiage to explain data widening. Hopefully this helps. this reference is now out of date.

* Working with regular indices
* Need to preserve distinct matches
* Need to match on any field in a lookup index

> * You use document or field level security

Reviewer: Or if we want to restrict users to specific lookup indices that they can use. (ENRICH is all or nothing.)

To use `LOOKUP JOIN`, you must have:

> * Data types of join key and join field in the lookup index need to generally be the same. As long as the field data type is structurally capable of handling the data that you are passing through. For example `short,byte` is considered equal to `integer`. Also, text fields can be used on the left hand side if and only if there is an exact subfield whose name is suffixed with `.keyword`.

Reviewer: Hmm, the wording with "structurally capable" is still quite abstract.

We could point to the list of [supported datatypes](https://www.elastic.co/guide/en/elasticsearch/reference/8.18/esql-limitations.html#_supported_types), this (very briefly) mentions which data types are all represented as `double` and `integer`, resp.

Maybe we can say that the data types need to be generally the same, resp. be represented as the same type as detailed in the supported data types list?

Reviewer 2: I agree this section isn't super clear: 

What about something like the following:

```markdown
## Prerequisites for LOOKUP JOIN

To use `LOOKUP JOIN`, the following requirements must be met:

* **Compatible data types**: The join key and join field in the lookup index must have compatible data types. This means:
   * The data types should generally be the same or be represented internally as the same type
   * Numeric types follow these compatibility rules:
     * `short` and `byte` are compatible with `integer` (all represented as `int`)
     * `float`, `half_float`, and `scaled_float` are compatible with `double` (all represented as `double`)
   * For text fields: You can use text fields on the left-hand side of the join only if they have a `.keyword` subfield 

For a complete list of supported data types and their internal representations, see the [Supported Field Types documentation](link).
```

> ## Limitations

Reviewer: Another limitation: only one join key field can be used for now, and we can only join on equality instead of more general conditions.

Reviewer: Sorry, this is already mentioned.
